File: ruby/game.rb
#!/usr/bin/env ruby

# A simple 2D position with + operator
class Position
  attr_reader :x, :y

  def initialize(x, y)
    @x, @y = x, y
  end

  def +(other)
    self.class.new(x + other.x, y + other.y)
  end

  def zero?
    x.zero? && y.zero?
  end
end

class GameOfLife
  ALIVE = "*"
  DEAD  = "."

  def initialize(rows = 20, cols = 90)
    @rows, @cols = rows, cols
    @board  = Array.new(rows) { Array.new(cols, DEAD) }
    @buffer = Array.new(rows) { Array.new(cols, DEAD) }
  end

  def run
    randomize_state
    loop do
      clear_screen
      draw_board
      next_state
      sleep 0.050
    end
  end

  private

  # ANSI clear + home
  def clear_screen
    print "\e[2J\e[H"
  end

  def randomize_state
    each_position do |pos|
      set_cell(ALIVE, pos) if rand(2).zero?
    end
  end

  def next_state
    each_position do |pos|
      live_n = neighbors(pos).count { |n| 
        get_cell(n) == ALIVE 
      }
      current = get_cell(pos)

      new_state =
        if current == ALIVE && (live_n == 2 || live_n == 3)
          ALIVE
        elsif current == DEAD && live_n == 3
          ALIVE
        else
          DEAD
        end

      @buffer[pos.y][pos.x] = new_state
    end

    # swap buffers
    @board, @buffer = @buffer, @board
  end

  def draw_board
    @rows.times do |y|
      @cols.times do |x|
        print get_cell(Position.new(x, y))
      end
      puts
    end
  end

  # ==== Helpers ====

  def get_cell(pos)
    @board[pos.y % @rows][pos.x % @cols]
  end

  def set_cell(state, pos)
    @board[pos.y % @rows][pos.x % @cols] = state
  end

  def each_position
    @rows.times do |y|
      @cols.times do |x|
        yield Position.new(x, y)
      end
    end
  end

  def neighbors(pos)
    deltas = [
      [-1,-1],[0,-1],[1,-1],
      [-1, 0],      [1, 0],
      [-1, 1],[0, 1],[1, 1],
    ]
    deltas
      .map { |dx, dy| pos + Position.new(dx, dy) }
  end
end

if __FILE__ == $0
  GameOfLife.new.run
end

File: swift/game.swift
import Foundation

final class GameOfLife {
    enum CellState: Character {
        case alive = "*"
        case dead = "."
    }

    struct Position {
        let x: Int
        let y: Int

        var isZero: Bool {
            x == 0 && y == 0
        }

        static func +(lhs: Self, rhs: Self) -> Self {
            .init(x: lhs.x + rhs.x, y: lhs.y + rhs.y)
        }
    }

    private let rows: Int
    private let cols: Int
    private var board: [CellState]
    private var buffer: [CellState]

    init(rows: Int = 20, cols: Int = 90) {
        self.rows = rows
        self.cols = cols
        self.board = Array(repeating: CellState.dead, count: rows*cols)
        self.buffer = Array(repeating: CellState.dead, count: rows*cols)
    }

    private func index(from position: Position) -> Int {
        let x = (position.x%cols + cols)%cols
        let y = (position.y%rows + rows)%rows
        return y*cols+x
    }

    func drawBoard() {
        clearScreen()
        for y in 0..<rows {
            for x in 0..<cols {
                let index = index(from: Position(x: x, y: y))
                print("\(board[index].rawValue)", terminator:"")
            }
            print("")
        }
    }

    private func countAliveNeighbors(at position: Position) -> Int {
        var aliveCount = 0
        for yOff in -1...1 {
            for xOff in -1...1 {
                let offset = Position(x: xOff, y: yOff)
                if offset.isZero { continue }
                aliveCount += getCell(at: position + offset) == .alive ? 1 : 0
            }
        }
        return aliveCount
    }

    private func setCell(state: CellState, at position: Position) {
        board[index(from: position)] = state;
    }

    private func getCell(at position: Position) -> CellState {
        board[index(from: position)]
    }

    private func nextState() {
        for y in 0..<rows {
            for x in 0..<cols {
                let position = Position(x: x, y: y)
                let count = countAliveNeighbors(at: position)
                var next: CellState = .dead
                if count == 3 || (getCell(at: position) == .alive && count == 2) {
                    next = .alive
                }
                buffer[index(from: position)] = next
            }
        }
        board = buffer
    }

    private func clearScreen() {
        // ANSI escape: clear screen and move cursor to home
        print("\u{1B}[2J\u{1B}[H", terminator: "")
    }

    private func randomizeState() {
        for y in 0..<rows {
            for x in 0..<cols {
                if Int.random(in: 0...1) == 1 {
                    setCell(state: .alive, at: .init(x: x, y: y))
                }
            }
        }
    }

    func run() {
        randomizeState()
        while true {
            drawBoard()
            nextState()
            Thread.sleep(forTimeInterval:0.050)
        }
    }
}

func main() {
    let game = GameOfLife()
    game.run()
}


main()
